trpcultivate_phenotypes/src/TripalCultivateValidator/ValidatorTraits/Headers.php

```diff
@@ -0,0 +1,201 @@
+        if (empty(trim($header[ $key ]))) {
+          throw new \Exception('Headers Trait requires the header key: ' . $key . ' to be have a value.');
+        }
+
+        // Type value is one of valid types.
+        if ($key == 'type' && !in_array($header[ $key ], $this->types)) {
+          $str_types = implode(', ', $this->types);
+          throw new \Exception('Headers Trait requires the header key: ' . $key . ' value to be one of [' . $str_types . '].');
+        }
+      }
+
+      // With the header type already verified to be one of the valid types, 
+      // push the header into the designated type context array.
+      $this->context[ $this->context_key ][ $header['type'] ][ $index ] = $header['name'];
+    }
+  }
+  
+  /**
+   * Get required headers.
+   * 
+   * @return array
+   *   All headers of type required, keyed by the index (order) from
+   *   the headers array and header name as the value.
+   * 
+   *   The key required in the context headers array set by the setter method.
+   *   NOTE: the headers array is zero-based index.
+   * 
+   * @throws \Exception
+   *  - If the 'headers' key does not exists in the context array
+   *    (ie. the headers element has NOT been set).
+   */
+  public function getRequiredHeaders() {
+    
+    $type_key = 'required';
+
+    if (array_key_exists($this->context_key, $this->context)) {
+      return $this->context[ $this->context_key ][ $type_key ];
+    }
+    else {
+      throw new \Exception('Cannot retrieve ' . $type_key . ' headers from the context array as one has not been set by setHeaders() method.');
+    }
+  }
+  
+  /**
+   * Get optional headers.
+   * 
+   * @return array
+   *   All headers of type optional, keyed by the index (order) from
+   *   the headers array and header name as the value.
+   * 
+   *   The key optional in the context headers array set by the setter method.
+   *   NOTE: the headers array is zero-based index.
+   * 
+   * @throws \Exception
+   *  - If the 'headers' key does not exists in the context array
+   *    (ie. the headers element has NOT been set).
+   */
+  public function getOptionalHeaders() {
+    
+    $type_key = 'optional';
+
+    if (array_key_exists($this->context_key, $this->context)) {
+      return $this->context[ $this->context_key ][ $type_key ];
+    }
+    else {
+      throw new \Exception('Cannot retrieve ' . $type_key . ' headers from the context array as one has not been set by setHeaders() method.');
+    }
+  }
+
+  /**
+   * Get headers of type(s).
+   * 
+   * @param array $types
+   *   A list of header types to get. 
+   *   Default to required and optional header types.
+   * 
+   * @return array
+   *   A list or combination of lists of headers of type
+   *   defined by the types parameter.
+   *   
+   *   Returned array is keyed by the index (order) from
+   *   the headers array and header name as the value.
+   *   NOTE: the headers array is zero-based index.
```

Comment:
This was not updated after the design change to merge the types into a single array.

```suggestion
   * @return array
   *   A single list of headers matching a type defined by the types parameter.
   *   For example, if the types parameter includes required and optional the the 
   *   resulting array will container all headers in order of index assuming there
   *   are only required and optional types supported.
   *
   *   The returned array is keyed by the index (column order) from
   *   the headers array and header name as the value.
   *   NOTE: the headers array is zero-based index.
```